aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -150,22 +150,21 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
 
 } // namespace
 
-std::tuple<Tensor&, Tensor&> topk_out_cuda(const Tensor& self,
-              int64_t k, int64_t dim, bool largest, bool sorted,
-              Tensor& values,
-              Tensor& indices) {
+TORCH_IMPL_FUNC(topk_out_cuda)
+  (const Tensor& self,
+   int64_t k, int64_t dim, bool largest, bool sorted,
+   const Tensor& values,
+   const Tensor& indices) {
   TensorArg topK_arg{values, "topK", 1}, indices_arg{indices, "indices", 2}, input_arg{self, "self", 3};
   checkAllSameGPU("topk_out_cuda", {topK_arg, indices_arg, input_arg});
-  TORCH_CHECK(self.dtype() == values.dtype(), "expected input to match values dtype");
   TORCH_CHECK(indices.dtype() == at::kLong, "expected indices to be of type ", at::kLong);
   dim = at::maybe_wrap_dim(dim, self);
+
   int numDims = self.dim();
   numDims = numDims == 0 ? 1 : numDims;
   TORCH_CHECK(numDims <= MAX_DIMS, "input tensor has too many dimensions");
-  TORCH_CHECK(dim >= 0 && dim < numDims, "dim not in range");
-
   int64_t sliceSize = self.dim() == 0 ? 1 : self.size(dim);
-  TORCH_CHECK(k >= 0 && k <= sliceSize, "k not in range for dimension");
+
   Tensor input = self.contiguous();
 
   // Build the output size, which is the dim being selected set to
```

Comment:
The logic here and below should be removed; structured kernels handle resizing for you

Reply:
Right, I think those can mostly be removed but I'm not sure about the `MAX_DIMS` check as that's not in Sorting.cpp where the Meta function is defined.

Reply:
I agree with you that `MAX_DIMS` can stay in the kernel. It looks like a kernel implementation detail and the semantic meaning of the kernel is that it should support arbitrary numbers of dims.